cpp/src/parquet/annotations-test.cc

```diff
@@ -0,0 +1,1482 @@
+  // underlying primitive type(s) it can be applied to
+
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::StringType(),
+                                          Type::BYTE_ARRAY);
+
+  ConfirmNoPrimitiveTypeApplicability(LogicalAnnotation::MapType());
+  ConfirmNoPrimitiveTypeApplicability(LogicalAnnotation::ListType());
+
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::EnumType(),
+                                          Type::BYTE_ARRAY);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::DateType(), Type::INT32);
+
+  ConfirmSinglePrimitiveTypeApplicability(
+      LogicalAnnotation::TimeType(true, LogicalAnnotation::TimeUnit::MilliSeconds),
+      Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(
+      LogicalAnnotation::TimeType(true, LogicalAnnotation::TimeUnit::MicroSeconds),
+      Type::INT64);
+  ConfirmSinglePrimitiveTypeApplicability(
+      LogicalAnnotation::TimeType(true, LogicalAnnotation::TimeUnit::NanoSeconds),
+      Type::INT64);
+  ConfirmSinglePrimitiveTypeApplicability(
+      LogicalAnnotation::TimestampType(true, LogicalAnnotation::TimeUnit::MilliSeconds),
+      Type::INT64);
+  ConfirmSinglePrimitiveTypeApplicability(
+      LogicalAnnotation::TimestampType(true, LogicalAnnotation::TimeUnit::MicroSeconds),
+      Type::INT64);
+  ConfirmSinglePrimitiveTypeApplicability(
+      LogicalAnnotation::TimestampType(true, LogicalAnnotation::TimeUnit::NanoSeconds),
+      Type::INT64);
+
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(8, false),
+                                          Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(16, false),
+                                          Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(32, false),
+                                          Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(64, false),
+                                          Type::INT64);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(8, true),
+                                          Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(16, true),
+                                          Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(32, true),
+                                          Type::INT32);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::IntType(64, true),
+                                          Type::INT64);
+
+  ConfirmAnyPrimitiveTypeApplicability(LogicalAnnotation::NullType());
+
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::JSONType(),
+                                          Type::BYTE_ARRAY);
+  ConfirmSinglePrimitiveTypeApplicability(LogicalAnnotation::BSONType(),
+                                          Type::BYTE_ARRAY);
```

Comment:
Fixed in commit f1e1b8a8eb8147d9e9f02bed3c6f2c9f5910fd44.  I used a struct rather than a pair for consistency with other tests in the file, but can change to a pair if you favor it strongly.